cmd/livepeer/latency_test.go

```diff
@@ -0,0 +1,334 @@
+	return time.Duration(seconds * 1000000000.0)
+}
+
+type SampleSegment struct {
+	Index      int
+	FrameCount int
+	Duration   time.Duration
+	Path       string
+}
+
+var sampleSegments = []SampleSegment{
+	{0, 360, 15000 * time.Millisecond, "./samples/16/sample_1_360_15000.ts"},
+	{1, 69, 2875 * time.Millisecond, "./samples/2/sample_0_69_2875.ts"},
+	{2, 53, 2208 * time.Millisecond, "./samples/2/sample_1_53_2208.ts"},
+	{3, 30, 1250 * time.Millisecond, "./samples/2/sample_2_30_1250.ts"},
+	{4, 409, 17041 * time.Millisecond, "./samples/16/sample_0_409_17041.ts"},
+}
+
+var manifestID = "13f63dab-5984-4580-b50d-7ab6753dc7cb"
+
+type ChunkLatency struct {
+	ByteCount int
+	Duration  time.Duration
+}
+
+type TransferLatency struct {
+	Chunks []ChunkLatency
+	Start  time.Time
+}
+
+func (l *TransferLatency) TotalTimeBytes() (bytes int, duration time.Duration) {
+	for _, chunk := range l.Chunks {
+		bytes += chunk.ByteCount
+		duration += chunk.Duration
+	}
+	return
+}
+
+func (l *TransferLatency) lerpToReceiveTimeline(t *testing.T, recvFraction float64, recvTimestamp time.Time) (milliseconds float64) {
+	totalSentBytes, _ := l.TotalTimeBytes()
+	sentTimestamp := l.Start
+	sentBytes := int(0)
+	for _, chunk := range l.Chunks {
+		nextBytes := sentBytes + chunk.ByteCount
+		sentFraction := float64(nextBytes) / float64(totalSentBytes)
+		if sentFraction >= recvFraction {
+			targetBytes := recvFraction * float64(totalSentBytes)
+			timeIncrement := (float64(chunk.Duration) / float64(time.Millisecond)) * ((targetBytes - float64(sentBytes)) / float64(chunk.ByteCount))
+			targetTimestamp := sentTimestamp.Add(time.Duration(timeIncrement * float64(time.Millisecond)))
+			return float64(recvTimestamp.Sub(targetTimestamp)) / float64(time.Millisecond)
+		}
+		sentTimestamp = sentTimestamp.Add(chunk.Duration)
+		sentBytes = nextBytes
+	}
+	assert.True(t, false, "Interpolate error")
+	return 0
+}
+
+type TranscodingLatencyCheckpoint struct {
+	Milliseconds float64
+	ByteCount    int
+}
+
+type TranscodingLatency struct {
+	Checkpoints []TranscodingLatencyCheckpoint
+}
+
+func calculateTranscodingLatency(t *testing.T, sending, receiving TransferLatency) (r TranscodingLatency) {
+	totalRecvBytes, _ := receiving.TotalTimeBytes()
+
+	recvTimestamp := receiving.Start
+	recvBytes := int(0)
+
+	for _, chunk := range receiving.Chunks {
+		recvTimestamp = recvTimestamp.Add(chunk.Duration)
+		recvBytes += chunk.ByteCount
+		recvFraction := float64(recvBytes) / float64(totalRecvBytes)
+		// `latency` should be calculated on frames. We base our calculation on data size.
+		// This is not precise as some frames are larger than others.
+		// Should be good for our test here.
+		latency := sending.lerpToReceiveTimeline(t, recvFraction, recvTimestamp)
+		r.Checkpoints = append(r.Checkpoints, TranscodingLatencyCheckpoint{
+			Milliseconds: latency,
+			ByteCount:    chunk.ByteCount,
+		})
+	}
+	return
+}
+
+func min(a, b int) int {
```

Comment:
You could use `math.MinInt` for this

Reply:
Also my initial thought, turns out [MinInt](https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/math/const.go;l=43) is minimal integer value constant :)

Reply:
Haha my bad!